### PR TITLE
Add support of multiple providers

### PR DESCRIPTION
**Pull request by asinyagin**
Hey,

I'd like to add support of multiple providers to web3. I was asking about workaround for it in #297. I've created this PR to discuss the idea, it changes only `eth.getBalance()`, so now the method can work through multiple providers.

I'll add comments to the diff to mark the main points.

### CODE REVIEW

File: lib/web3.js [hunk 1]
```diff
@@ -78,56 +79,60 @@ var setupProperties = function (obj, properties) {
 };
 
 /// setups web3 object, and it's in-browser executed methods
-var web3 = {};
-web3.providers = {};
-web3.currentProvider = null;
-web3.version = {};
-web3.version.api = version.version;
-web3.eth = {};
+function Web3(provider) {
```
Comment (asinyagin) [inline lib/web3.js]: That's a constructor for `web3` object. The object stores links to `requestManager` and namespaces like `eth` (`eth1` in this case).

File: lib/web3/method.js [hunk 2]
```diff
@@ -168,5 +168,30 @@ Method.prototype.send = function () {
     return this.formatOutput(RequestManager.getInstance().send(payload));
 };
 
-module.exports = Method;
+Method.prototype.attachToObject1 = function (obj) {
```
Comment (asinyagin) [inline lib/web3/method.js]: It uses closure to attach method to namespace.

File: lib/web3/method.js [hunk 3]
```diff
@@ -168,5 +168,30 @@ Method.prototype.send = function () {
...
+        obj[name[0]] = func; 
+    }
+};
+
+Method.prototype.buildCall = function() {
+    var method = this;
+    return function send() {
+        var payload = method.toPayload(Array.prototype.slice.call(arguments));
+        if (payload.callback) {
+            return this.web3.requestManager.sendAsync(payload, function (err, result) {
```
Comment (asinyagin) [inline lib/web3/method.js]: It sends a request through namespace's `requestManager`.

File: lib/web3/methods/eth1.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
+var formatters = require('../formatters');
+var utils = require('../../utils/utils');
+var Method = require('../method');
+var Property = require('../property');
+var c = require('../../utils/config');
+
+function Eth(web3) {
```
Comment (asinyagin) [inline lib/web3/methods/eth1.js]: Here's the way to define a namespace with methods and properties.

File: lib/web3.js (hunk 1 above)
Comment (frozeman) [inline lib/web3.js]: So you need to instantiate `Web3`?
So you would like to have multiple instances, with different providers running in the same app?

web3 is designed to be an object like the `chrome` or later `mist` object. It gives "native" functions and the dapp shouldn't even care how it connects.

File: lib/web3/methods/eth1.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+var methods = [
+    new Method({
+        name: 'getBalance',
+        call: 'eth_getBalance',
+        params: 2,
+        inputFormatter: [formatters.inputAddressFormatter, formatters.inputDefaultBlockNumberFormatter],
+        outputFormatter: formatters.outputBigNumberFormatter
+    })
+];
+methods.forEach(function(method) { method.attachToObject1(Eth.prototype) });
```
Comment (frozeman) [inline lib/web3/methods/eth1.js]: I like attaching it here, rather than in `web3.js`. @debris any opinion?

File: lib/web3.js (hunk 1 above)
Comment (asinyagin) [inline lib/web3.js]: Yeah, our ide works with 2 and more JSON RPC in parallel.

Okay, I see the problem. Yeah, that could be unnecessary complication. But if you think it's useful to work with multiple providers, you could get rid of `RequestManager` singleton and current global `web3` and define global object like `window.web3 = new Web3()`.

**Comment by debris** [comment]
I like this pr a lot. I was thinking about this kind of refactor for some time :+1: 

And after we finish that, I would like to refactor even further. I would like to have strongly typed input params. I even did a small [demo](https://github.com/debris/web3.ts) in typescript. We could have similar types to [these](https://github.com/debris/web3.ts/blob/develop/lib/types.ts). It would also solve issues similar to #322. What do you think about this? @frozeman @asinyagin

**Comment by debris** [comment]
Of course we would to this in 2 (or more) separate steps

**Comment by asinyagin** [comment]
@debris nice! Looking forward to test the version with multiple providers. And I agree it's a good idea to validate args. I have such validators to check params before giving them to web3.

**Comment by frozeman** [comment]
How would an example look like for the strict params? Would you simply throw an error if they don't match?

**Comment by asinyagin** [comment]
I'm not sure what's a good way in general, but in my case validator (for uintN from Solidity) returns an array with errors:
```javascript
define(function() {
    return function(type) {
        var size = parseInt(type.substr(3));
        return {
            validate: function(value) {
                var errors = [];
                if (!value.match(/^-?\d+$/))
                    errors.push('int must contain only sign and digits.');
                else {
                    var max = new BigNumber(2).pow(size).minus(1),
                        val = new BigNumber(value, 10);
                    if (!val.lessThan(max))
                        errors.push(type + ' must be less than 2^' + size + '.');
                }
                return errors;
            }
        };
    };
});
```

**Pull request closed without merging**